src/optimizer/rule_impls.cpp

```diff
@@ -14,64 +14,80 @@
 #include "optimizer/util.h"
 #include "optimizer/operators.h"
 #include "storage/data_table.h"
+#include "optimizer/properties.h"
 
 #include <memory>
+#include <include/optimizer/optimizer_metadata.h>
 
 namespace peloton {
 namespace optimizer {
 
+//===--------------------------------------------------------------------===//
+// Transformation rules
+//===--------------------------------------------------------------------===//
+
 ///////////////////////////////////////////////////////////////////////////////
 /// InnerJoinCommutativity
 InnerJoinCommutativity::InnerJoinCommutativity() {
-  logical = true;
+  type_ = RuleType::INNER_JOIN_COMMUTE;
 
   std::shared_ptr<Pattern> left_child(std::make_shared<Pattern>(OpType::Leaf));
   std::shared_ptr<Pattern> right_child(std::make_shared<Pattern>(OpType::Leaf));
-  std::shared_ptr<Pattern> predicate(std::make_shared<Pattern>(OpType::Leaf));
   match_pattern = std::make_shared<Pattern>(OpType::InnerJoin);
   match_pattern->AddChild(left_child);
   match_pattern->AddChild(right_child);
-  match_pattern->AddChild(predicate);
 }
 
 bool InnerJoinCommutativity::Check(std::shared_ptr<OperatorExpression> expr,
-                                   Memo *memo) const {
-  (void)memo;
+                                   OptimizeContext *context) const {
+  (void)context;
   (void)expr;
   return true;
 }
 
 void InnerJoinCommutativity::Transform(
     std::shared_ptr<OperatorExpression> input,
-    std::vector<std::shared_ptr<OperatorExpression>> &transformed) const {
-  auto result_plan =
-      std::make_shared<OperatorExpression>(LogicalInnerJoin::make());
+    std::vector<std::shared_ptr<OperatorExpression>> &transformed,
+    UNUSED_ATTRIBUTE OptimizeContext *context) const {
+  auto join_op = input->Op().As<LogicalInnerJoin>();
+  auto join_predicates =
+      std::vector<AnnotatedExpression>(join_op->join_predicates);
+  auto result_plan = std::make_shared<OperatorExpression>(
+      LogicalInnerJoin::make(join_predicates));
   std::vector<std::shared_ptr<OperatorExpression>> children = input->Children();
   PL_ASSERT(children.size() == 2);
+  LOG_TRACE(
+      "Reorder left child with op %s and right child with op %s for inner join",
+      children[0]->Op().name().c_str(), children[1]->Op().name().c_str());
   result_plan->PushChild(children[1]);
   result_plan->PushChild(children[0]);
 
   transformed.push_back(result_plan);
```

Comment:
This is just a rule to transform (A join B) to (B join A). We need to support transform ((A join B) join C) to (A join (B join C)) and the opposite way to fully support join order enumeration.

Reply:
@chenboy Is this something that we can add in January after we get this PR in? You can work with @GustavoAngulo on it when get gets back.

Reply:
Sure. This could be implemented soon. I think it would be better if we get the statistics stuff in first. We can write better test cases to see if the optimizer provide the optimal plan.

Reply:
I agree that we could get the statistics stuff in first.